### PR TITLE
feat(kinesisfirehose): support Redshift table destination

### LINKED ISSUE #15503: (aws-kinesisfirehose): Redshift destination for DeliveryStream
Allow specifying a Redshift table as a destination for `DeliveryStream`.

Part of https://github.com/aws/aws-cdk/issues/7536

### Use Case

Specifying a Redshift table destination in CloudFormation or with L1 constructs is very complex, this should be included in the L2 to make it easier for users.

### Proposed Solution

See [RFC 340: Kinesis Firehose Delivery Stream](https://github.com/aws/aws-cdk-rfcs/pull/342)

And [this branch with the prototype](https://github.com/madeline-k/aws-cdk/tree/madeline-k/feat/kinesisfirehose-l2)

* [ ] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by BenChaimberg**
Kinesis Firehose supports Redshift tables as destinations for delivery streams. This feature provides an L2 construct for this type of destination.

This feature includes the following improvements for users over configuring the destination manually:
- integrates with `aws-redshift` module
  - leverages `IUser` and `ITable` to allow users to one-click deploy a delivery stream without needing to manually create resources within a Redshift cluster
  - automatically creates user ("assumed" by Firehose to deliver data) and grants minimally necessary permissions on table
- validates and adds necessary cluster configuration to ensure Firehose can access the cluster
- automatically creates intermediary resources (S3 bucket, IAM role) needed for Firehose to delivery data to Redshift
- other generic delivery stream destination features that were created previously (see: [readme](https://docs.aws.amazon.com/cdk/api/latest/docs/aws-kinesisfirehose-readme.html))

closes #15503

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-kinesisfirehose-destinations/lib/redshift-cluster.ts [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+   * Username for user that has permission to insert records into a Redshift table.
+   */
+  readonly username: string;
+
+  /**
+   * Password for user that has permission to insert records into a Redshift table.
+   *
+   * Do not put passwords in your CDK code directly.
+   *
+   * @default - a Secrets Manager generated password.
```
Comment (madeline-k) [inline packages/@aws-cdk/aws-kinesisfirehose-destinations/lib/redshift-cluster.ts:28]: How would the generated password work? This might become clear later, but will you create a new Redshift user with the right permissions in CDK?

File: packages/@aws-cdk/aws-kinesisfirehose-destinations/lib/redshift-cluster.ts [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+   */
+  readonly password?: cdk.SecretValue;
+
+  /**
+   * KMS key to encrypt the generated secret.
```
Comment (madeline-k) [inline packages/@aws-cdk/aws-kinesisfirehose-destinations/lib/redshift-cluster.ts:33]: Is this referring to the secret that will be generated to store the Redshift credentials (username and password)?

Should we also allow customers to just pass in a SecretsManager secret that already contains their Redshift credentials?

File: packages/@aws-cdk/aws-kinesisfirehose-destinations/lib/redshift-cluster.ts [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+  readonly tableColumns: RedshiftColumn[];
+
+  /**
+   * The secret that holds Redshift cluster credentials for a user with administrator privileges.
+   *
+   * Used to create the user that Firehose assumes and the table that data is inserted into.
+   *
+   * @default - the master secret is taken from the cluster.
+   */
+  readonly masterSecret?: secretsmanager.ISecret;
```
Comment (madeline-k) [inline packages/@aws-cdk/aws-kinesisfirehose-destinations/lib/redshift-cluster.ts:86]: If this is specified, why would the user also need to specify a 'RedshiftUser'? If they specify a secret that is the "Credentials for a Redshift cluster" type, then would the username and password and encryption key all be there as part of the secret? 

Can we enforce that the secret passed in here follows the "Credentials for a Redshift cluster" format? This is defined in the AWS console, but does not seem to be defined in CFN.

We should choose an inclusive name for this term: `adminSecret`, `superUserCredentials`?

File: packages/@aws-cdk/aws-kinesisfirehose-destinations/lib/redshift-cluster.ts [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+
+  /**
+   * Parameters given to the COPY command that is used to move data from S3 to Redshift.
+   *
+   * @see https://docs.aws.amazon.com/firehose/latest/APIReference/API_CopyCommand.html
+   *
+   * @default '' - no extra parameters are provided to the Redshift COPY command
+   */
+  readonly copyOptions?: string;
```
Comment (madeline-k) [inline packages/@aws-cdk/aws-kinesisfirehose-destinations/lib/redshift-cluster.ts:95]: In what scenario would a customer actually want to hand-write this? I think it should always be able to be generated from the s3 bucket and table columns, and redshift table info.

File: packages/@aws-cdk/aws-kinesisfirehose-destinations/lib/redshift-cluster.ts [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+
+  /**
+   * The length of time during which Firehose retries delivery after a failure.
```
Comment (madeline-k) [inline packages/@aws-cdk/aws-kinesisfirehose-destinations/lib/redshift-cluster.ts:98]: Is there a min and max for this?

File: packages/@aws-cdk/aws-kinesisfirehose-destinations/lib/redshift-cluster.ts [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+   *
+   * @default - a role will be created for you.
+   */
+  readonly bucketAccessRole?: iam.IRole;
+}
+
+/**
+ * Redshift delivery stream destination.
+ */
+export class RedshiftCluster implements firehose.IDestination {
```
Comment (madeline-k) [inline packages/@aws-cdk/aws-kinesisfirehose-destinations/lib/redshift-cluster.ts:124]: I think this should be called `RedshiftTable`. The destination has to be specific to a table, not just a cluster in general.

File: packages/@aws-cdk/aws-kinesisfirehose-destinations/lib/redshift-cluster.ts [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+  /**
+   * The secret Firehose will use to access the Redshift cluster.
+   */
+  public secret?: secretsmanager.ISecret;
+
+  private masterSecret: secretsmanager.ISecret;
+
+  constructor(private readonly cluster: redshift.ICluster, private readonly props: RedshiftDestinationProps) {
+    if (!cluster.publiclyAccessible) {
+      throw new Error('Redshift cluster used as Firehose destination must be publicly accessible');
```
Comment (madeline-k) [inline packages/@aws-cdk/aws-kinesisfirehose-destinations/lib/redshift-cluster.ts:134]:
```suggestion
      throw new Error('Redshift cluster used as delivery stream destination is not publicly accessible');
```

File: packages/@aws-cdk/aws-kinesisfirehose-destinations/lib/redshift-cluster.ts [hunk 8]
```diff
@@ -0,0 +1,276 @@
...
+    }
+    if (!cluster.subnetGroup?.selectedSubnets?.hasPublic) {
+      throw new Error('Redshift cluster used as Firehose destination must be located in a public subnet');
```
Comment (madeline-k) [inline packages/@aws-cdk/aws-kinesisfirehose-destinations/lib/redshift-cluster.ts:137]:
```suggestion
      throw new Error('Redshift cluster used as delivery stream destination is not located in a public subnet');
```

File: packages/@aws-cdk/aws-kinesisfirehose-destinations/lib/redshift-cluster.ts [hunk 9]
```diff
@@ -0,0 +1,276 @@
...
+    }
+    const masterSecret = props.masterSecret ?? (cluster instanceof redshift.Cluster ? cluster.secret : undefined);
+    if (!masterSecret) {
+      throw new Error('Master secret must be provided or Redshift cluster must generate a master secret');
```
Comment (madeline-k) [inline packages/@aws-cdk/aws-kinesisfirehose-destinations/lib/redshift-cluster.ts:141]: I think this error is confusing. How would a user remediate this? In what scenarios would the secret be undefined at this point?

File: packages/@aws-cdk/aws-kinesisfirehose-destinations/lib/redshift-cluster.ts [hunk 10]
```diff
@@ -0,0 +1,276 @@
...
+    }
+    this.masterSecret = masterSecret;
+    if (props.retryTimeout && props.retryTimeout.toSeconds() > cdk.Duration.hours(2).toSeconds()) {
+      throw new Error('Retry timeout must be less that 2 hours');
```
Comment (madeline-k) [inline packages/@aws-cdk/aws-kinesisfirehose-destinations/lib/redshift-cluster.ts:145]: Include the customer-provided value in the error message.

File: packages/@aws-cdk/aws-kinesisfirehose-destinations/lib/redshift-cluster.ts [hunk 11]
```diff
@@ -0,0 +1,276 @@
...
+    }
+    if (props.compression === Compression.SNAPPY || props.compression === Compression.HADOOP_SNAPPY || props.compression === Compression.ZIP) {
+      throw new Error('Compression must not be HADOOP_SNAPPY, SNAPPY, or ZIP');
```
Comment (madeline-k) [inline packages/@aws-cdk/aws-kinesisfirehose-destinations/lib/redshift-cluster.ts:148]:
```suggestion
      throw new Error(`${props.Compression} compression is not supported by Redshift destinations. Choose one of: _______`);
```

File: packages/@aws-cdk/aws-kinesisfirehose-destinations/lib/redshift-cluster.ts [hunk 12]
```diff
@@ -0,0 +1,276 @@
...
+    }
+    if (props.s3Backup?.mode === BackupMode.FAILED) {
+      throw new Error('Redshift delivery stream destination only supports ALL and DISABLED BackupMode, given FAILED');
```
Comment (madeline-k) [inline packages/@aws-cdk/aws-kinesisfirehose-destinations/lib/redshift-cluster.ts:151]: DISABLED backup mode doesn't exist anymore in the CDK API. 
```suggestion
      throw new Error('Redshift delivery stream destination does not support FAILED BackupMode.');
```

File: packages/@aws-cdk/aws-kinesisfirehose-destinations/lib/redshift-cluster.ts [hunk 13]
```diff
@@ -0,0 +1,276 @@
...
+  }
+
+  private createRetryOptions(retryTimeout?: cdk.Duration): firehose.CfnDeliveryStream.RedshiftRetryOptionsProperty | undefined {
+    return retryTimeout ? {
+      durationInSeconds: retryTimeout.toSeconds(),
+    } : undefined;
+  }
+
+  private getS3BackupMode(): string | undefined {
+    return (this.props.s3Backup?.bucket && !this.props.s3Backup.mode) || this.props.s3Backup?.mode === BackupMode.ALL
```
Comment (madeline-k) [inline packages/@aws-cdk/aws-kinesisfirehose-destinations/lib/redshift-cluster.ts:272]: This almost the same logic as: https://github.com/aws/aws-cdk/blob/master/packages/@aws-cdk/aws-kinesisfirehose-destinations/lib/s3-bucket.ts#L58 And I think they end up being equivalent. 

Should this be shared somewhere?

File: packages/@aws-cdk/aws-kinesisfirehose-destinations/lib/redshift/table.ts [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+  readonly tableName: string;
+
+  /**
+   * The table columns that the source fields will be loaded into.
+   */
+  readonly tableColumns: RedshiftColumn[];
+}
+
+/**
+ * Custom resource to create the Redshift table where Firehose will load data.
```
Comment (madeline-k) [inline packages/@aws-cdk/aws-kinesisfirehose-destinations/lib/redshift/table.ts:46]: oooh interesting. I wonder if people will actually use this, or if they will use a Redshift table that they've already created? 

I would say this feature should go in a separate PR, but I guess you need it for integration tests.

File: packages/@aws-cdk/aws-kinesisfirehose-destinations/lib/redshift/table.ts [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+   */
+  readonly masterSecret: secretsmanager.ISecret;
+
+  /**
+   * The database containing the desired output table.
+   */
+  readonly database: string;
+
+  /**
+   * The table that data should be inserted into.
```
Comment (madeline-k) [inline packages/@aws-cdk/aws-kinesisfirehose-destinations/lib/redshift/table.ts:35]: What happens if the table already exists? I think this comment should also describe that if the table doesn't already exist, it will be created.

File: packages/@aws-cdk/aws-kinesisfirehose-destinations/lib/redshift/table.ts [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+  /**
+   * Name of the table.
+   */
+  public readonly tableName: string;
+
+  constructor(scope: Construct, id: string, props: FirehoseRedshiftTableProps) {
+    super(scope, id);
+
+    const handler = new lambda.Function(this, 'Handler', {
+      code: lambda.Code.fromAsset(path.join(__dirname, 'provider')),
```
Comment (madeline-k) [inline packages/@aws-cdk/aws-kinesisfirehose-destinations/lib/redshift/table.ts:58]: Can you use a name more specific than 'provider' here? I don't have any specific suggestions, but the directory name should make it more clear what all this lambda code is for.

Or is this a pattern for custom resources that are `Provider`s?

**Review by madeline-k — COMMENTED** [review]
First pass comments. 

This is doing a lot of heavy-lifting for the user in creating the Redshift table, user credentials, etc. Which is great and I think that's eventually what we want for this feature. But would it be possible to separate out some of these things into other PRs? And for this first PR keep the CDK experience closer to the AWS Console experience? Understand this might make it hard to test. I am trying to think of ways to make this shorter and easier to review.

File: packages/@aws-cdk/aws-kinesisfirehose-destinations/lib/redshift-cluster.ts (hunk 4 above)
Comment (madeline-k) [reply packages/@aws-cdk/aws-kinesisfirehose-destinations/lib/redshift-cluster.ts:95]: Sorry just realized, this is not the COPY command itself, just the options. Disregard :)

File: packages/@aws-cdk/aws-kinesisfirehose-destinations/lib/redshift-cluster.ts [hunk 17]
```diff
@@ -0,0 +1,276 @@
...
+ * The Redshift user Firehose will assume to deliver data to Redshift
+ */
+export interface RedshiftUser {
+  /**
+   * Username for user that has permission to insert records into a Redshift table.
+   */
+  readonly username: string;
+
+  /**
+   * Password for user that has permission to insert records into a Redshift table.
```
[the remark this answers is not in the document]
Comment (BenChaimberg) [reply packages/@aws-cdk/aws-kinesisfirehose-destinations/lib/redshift-cluster.ts:24]:
```suggestion
   * Password for user that has permission to insert records into the destination Redshift table.
```

File: packages/@aws-cdk/aws-kinesisfirehose-destinations/lib/redshift-cluster.ts (hunk 13 above)
Comment (madeline-k) [reply packages/@aws-cdk/aws-kinesisfirehose-destinations/lib/redshift-cluster.ts]: That sounds good to me. A little bit of duplication is okay here in order to make it clear that each destination has full control over this part. The only way I can think of to make it clear and not have duplication would be to have a function called something like `getS3BackupModeForS3OrRedshift` which seems clunky and not flexible

File: packages/@aws-cdk/aws-kinesisfirehose-destinations/lib/redshift/table.ts [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+  /**
+   * The table columns that the source fields will be loaded into.
+   */
+  readonly tableColumns: RedshiftColumn[];
+}
+
+/**
+ * Custom resource to create the Redshift table where Firehose will load data.
+ */
+export class FirehoseRedshiftTable extends CoreConstruct {
```
Comment (madeline-k) [inline packages/@aws-cdk/aws-kinesisfirehose-destinations/lib/redshift/table.ts:43]: I am struggling to come up with a better name for this and `FirehoseRedshiftUser`. I think there will be some confusion with the following names:

1. `FirehoseRedshiftTable` - custom resource to create the table if needed
2. `FirehoseRedshiftUser` - custom resource to create the user if needed
3. `RedshiftTable` - the actual delivery stream destination
4. `RedshiftUser` - the class representing a user that a customer will pass into the constructor props

The relationship between 1 and 3 is not the same as the relationship between 2 and 4, even though the names have the same relationship. 

I think I don't like the prefix 'Firehose' for 1&2 since it seems to imply that Firehose has some ownership of the Table and User. 1&2 are really redshift database resources created by the CDK for Firehose to use. 

This is a very roundabout way of saying: I think 1&2 should be called something like `GeneratedRedshiftX` `CreatedRedshiftX`, `RedshiftXGenerator`, `RedshiftXBuilder`. I am also open to a different model here.

File: packages/@aws-cdk/aws-kinesisfirehose-destinations/lib/redshift-table.ts [hunk 19]
```diff
@@ -0,0 +1,311 @@
...
+
+  /**
+   * Generate credentials and create a Redshift user.
+   *
+   * The user will be given INSERT permissions to the destination table. The credentials for the user will be stored in AWS Secrets Manager.
+   *
+   * @param encryptionKey KMS key to encrypt the generated secret. If not provided, the default AWS managed key is used.
+   */
+  static create(encryptionKey?: kms.IKey): RedshiftUser {
+    return new RedshiftUser(undefined, undefined, encryptionKey);
```
[the remark this answers is not in the document]
Comment (BenChaimberg) [reply packages/@aws-cdk/aws-kinesisfirehose-destinations/lib/redshift-table.ts:38]: `new`? so it's a noun instead of a verb?

File: packages/@aws-cdk/aws-kinesisfirehose-destinations/lib/redshift-table.ts [hunk 20]
```diff
@@ -0,0 +1,311 @@
...
+      throw new Error('Redshift delivery stream destination does not support ${props.s3Backup?.mode} backup mode');
+    }
+
+    const generateTable = !props.tableName;
+    const generateUser = !props.user.username || !props.user.password;
+    const adminUser = props.adminUser ?? (cluster instanceof redshift.Cluster ? cluster.secret : undefined);
+    if (!adminUser && (generateTable || generateUser)) {
+      throw new Error(
+        'Administrative access to the cluster is required but an admin user secret was not provided and either the Redshift cluster did not generate admin user credentials (they were provided explicitly) or the cluster was imported',
+      );
```
Comment (madeline-k) [inline packages/@aws-cdk/aws-kinesisfirehose-destinations/lib/redshift-table.ts:182]: This is a doozy of an error message. But I think you get the point across. Maybe consider splitting it up? This suggestion is not completely correct, but hopefully you get the idea. 

```suggestion
if (!(cluster instanceof redshift.Cluster)) {
      throw new Error(
        'Administrative access to the cluster is required but an admin user secret was not provided and the cluster was imported',
      );
}
if (!cluster.secret) {
      throw new Error(
        'Administrative access to the cluster is required but an admin user secret was not provided and the Redshift cluster did not generate admin user credentials (they were provided explicitly)',
      );
}  
```

File: packages/@aws-cdk/aws-kinesisfirehose-destinations/lib/redshift-table.ts [hunk 21]
```diff
@@ -0,0 +1,311 @@
...
+}
+
+/**
+ * Redshift table delivery stream destination.
+ */
+export class RedshiftTable implements firehose.IDestination {
+  /**
+   * The Redshift database that contains the table were data will be delivered.
+   */
+  public readonly database: string;
```
[the remark this answers is not in the document]
Comment (BenChaimberg) [reply packages/@aws-cdk/aws-kinesisfirehose-destinations/lib/redshift-table.ts:144]: Correct – Redshift doesn't have L2s for resources within a cluster. Deleted since unused

File: packages/@aws-cdk/aws-kinesisfirehose/README.md [hunk 22]
```diff
@@ -138,6 +138,54 @@ const s3Destination = new destinations.S3Bucket(bucket, {
...
+### Redshift
+
+A delivery stream can deliver data to a table within a Redshift cluster, using an
+intermediate S3 bucket and executing a Redshift `COPY` command. Redshift clusters must be
+placed in public subnets within an VPC, must be marked as publicly accessible, and cannot
+provide a master user password (it must be generated by the CDK). A Redshift user will be
+created within the cluster for the exclusive use of the delivery stream, and a Redshift
+table with the provided schema will be created within the provided database.
+
+```ts
```
[the remark this answers is not in the document]
Comment (BenChaimberg) [reply packages/@aws-cdk/aws-kinesisfirehose/README.md:150]: definitely out of date now – will fix

File: packages/@aws-cdk/aws-kinesisfirehose-destinations/lib/redshift-table.ts [hunk 23]
```diff
@@ -0,0 +1,311 @@
...
+export interface RedshiftColumn {
+  /**
+   * The name of the column.
+   */
+  readonly name: string;
+
+  /**
+   * The data type of the column.
+   */
+  readonly dataType: string;
```
[the remark this answers is not in the document]
Comment (BenChaimberg) [reply packages/@aws-cdk/aws-kinesisfirehose-destinations/lib/redshift-table.ts]: p2 IMO but yes would certainly be nice

File: packages/@aws-cdk/aws-kinesisfirehose-destinations/lib/redshift-table.ts [hunk 24]
```diff
@@ -0,0 +1,177 @@
...
+    } = this.props;
+    const cluster = this.table.cluster;
+
+    const role = this.props.role ?? new iam.Role(scope, 'Redshift Destination Role', {
+      assumedBy: new iam.ServicePrincipal('firehose.amazonaws.com'),
+    });
+
+    const endpoint = cluster.clusterEndpoint;
+    const jdbcUrl = `jdbc:redshift://${endpoint.hostname}:${cdk.Token.asString(endpoint.port)}/${this.table.databaseName}`;
+    cluster.connections.allowDefaultPortFrom(options.connections, 'Allow incoming connections from Kinesis Data Firehose');
```
Comment (BenChaimberg) [inline packages/@aws-cdk/aws-kinesisfirehose-destinations/lib/redshift-table.ts:104]: add test

File: packages/@aws-cdk/aws-kinesisfirehose-destinations/lib/redshift-table.ts [hunk 25]
```diff
@@ -0,0 +1,177 @@
...
+   * @default - a user is generated
+   */
+  readonly user?: redshift.IUser;
+
+  /**
+   * Parameters given to the COPY command that is used to move data from S3 to Redshift.
+   *
+   * @see https://docs.aws.amazon.com/firehose/latest/APIReference/API_CopyCommand.html
+   *
+   * @default '' - no extra parameters are provided to the Redshift COPY command
```
[the remark this answers is not in the document]
Comment (BenChaimberg) [reply packages/@aws-cdk/aws-kinesisfirehose-destinations/lib/redshift-table.ts:26]: > By default, the COPY command expects the source data to be in character-delimited UTF-8 text files. The default delimiter is a pipe character (`|`).

File: packages/@aws-cdk/aws-kinesisfirehose-destinations/lib/redshift-table.ts [hunk 26]
```diff
@@ -0,0 +1,177 @@
...
+
+    const endpoint = cluster.clusterEndpoint;
+    const jdbcUrl = `jdbc:redshift://${endpoint.hostname}:${cdk.Token.asString(endpoint.port)}/${this.table.databaseName}`;
+    cluster.connections.allowDefaultPortFrom(options.connections, 'Allow incoming connections from Kinesis Data Firehose');
+
+    const user = this.props.user ?? new redshift.User(scope, 'Redshift User', {
+      cluster: cluster,
+      databaseName: this.table.databaseName,
+    });
+    this.table.grant(user, redshift.TableAction.INSERT);
```
Comment (BenChaimberg) [inline packages/@aws-cdk/aws-kinesisfirehose-destinations/lib/redshift-table.ts:110]: the grant created by this needs to be added as a dependable, [below](https://github.com/aws/aws-cdk/pull/15837/files#diff-7ad3065138c0bcd78070f7f716e24f9f7e66f648475716bddf7803449799de88R162)

File: packages/@aws-cdk/aws-kinesisfirehose-destinations/lib/redshift-table.ts (hunk 25 above)
[the remark this answers is not in the document]
Comment (BenChaimberg) [reply packages/@aws-cdk/aws-kinesisfirehose-destinations/lib/redshift-table.ts:26]: don't think it's necessary to add a specific property defining the data format for the first release of this construct. However, I think we should document the default behaviour (that the COPY command assumes the data is text) somewhere and I'm not sure the best place to put that. @madeline-k any thoughts?

[the remark this answers is not in the document]
Comment (BenChaimberg) [reply packages/@aws-cdk/aws-kinesisfirehose-destinations/lib/redshift-table.ts:27]: I added it to the construct as a whole. I would have loved to add it on the `table` parameter in the construct's constructor, but it seems like our docs don't consistently render anything after the first sentence for `@param` annotations

**Comment by TheRealAmazonKendra** [comment]
This PR has been deemed to be abandoned, and will be automatically closed. Please create a new PR for these changes if you think this decision has been made in error.
